pkg/api/latest/latest.go

```diff
@@ -18,125 +18,108 @@ package latest
 
 import (
 	"fmt"
-	"sort"
-	"strings"
 
+	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/api/meta"
 	"k8s.io/kubernetes/pkg/api/registered"
 	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/api/v1"
 	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
 )
 
-var (
-	allGroups = GroupMetaMap{}
-	// Group is a shortcut to allGroups.Group.
-	Group = allGroups.Group
-	// RegisterGroup is a shortcut to allGroups.RegisterGroup.
-	RegisterGroup = allGroups.RegisterGroup
-	// GroupOrDie is a shortcut to allGroups.GroupOrDie.
-	GroupOrDie = allGroups.GroupOrDie
-	// AllPreferredGroupVersions returns the preferred versions of all
-	// registered groups in the form of "group1/version1,group2/version2,..."
-	AllPreferredGroupVersions = allGroups.AllPreferredGroupVersions
-	// IsRegistered is a shortcut to allGroups.IsRegistered.
-	IsRegistered = allGroups.IsRegistered
-)
+// ExternalVersions lists all allowed external versions for this group from most preferred to least preferred
+var ExternalVersions = []unversioned.GroupVersion{}
 
-// GroupMetaMap is a map between group names and their metadata.
-type GroupMetaMap map[string]*GroupMeta
+// PreferredExternalVersion is the most preferred external version (ExternalVersions[0])
+var PreferredExternalVersion = unversioned.GroupVersion{}
 
-// RegisterGroup registers a group to GroupMetaMap.
-func (g GroupMetaMap) RegisterGroup(group string) (*GroupMeta, error) {
-	_, found := g[group]
-	if found {
-		return nil, fmt.Errorf("group %v is already registered", g)
-	}
-	if len(registered.GroupVersionsForGroup(group)) == 0 {
-		return nil, fmt.Errorf("No version is registered for group %v", group)
+// Codec is the Codec for the most preferred ExternalVersion
+var Codec runtime.Codec
+
+// RESTMapper is a RESTMapper for all versions of the API group
+var RESTMapper meta.RESTMapper
+
+var Accessor = meta.NewAccessor()
+
+// availableVersions lists all known external versions for this group from most preferred to least preferred
+var availableVersions = []unversioned.GroupVersion{v1.SchemeGroupVersion}
+
+func init() {
+	finalExternalVersions := []unversioned.GroupVersion{}
+
+	for _, allowedVersion := range registered.GroupVersionsForGroup(api.SchemeGroupVersion.Group) {
+		for _, externalVersion := range availableVersions {
```

Comment:
> die if we encounter a registered version for this group that isn't in availableVersions, or leave as-is for now?

As-is for now until we sort out the future of `registered`.  I suspect it will fall out and this is already an advanced case.